subprojects/core/src/integTest/groovy/org/gradle/api/file/FileCollectionSymlinkIntegrationTest.groovy

```diff
@@ -174,6 +177,213 @@ class FileCollectionSymlinkIntegrationTest extends AbstractIntegrationSpec {
+        inputDirectoryWithBrokenLink.file('brokenInputFile').createLink(inputFileTarget)
+        def output = file("output.txt")
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+                @InputDirectory File inputDirectoryWithBrokenLink
+
+                @OutputFile File output
+
+                @TaskAction execute() {
+                    output.text = inputDirectoryWithBrokenLink.list()
+                }
+            }
+            task inputBrokenLinkNameCollector(type: CustomTask) {
+                inputDirectoryWithBrokenLink = file '${inputDirectoryWithBrokenLink}'
+                output = file '${output}'
+            }
+        """
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "[brokenInputFile]"
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+
+        when:
+        inputFileTarget.createFile()
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+    }
+
+    // Validation fails with broken links assigned directly to `@InputFile`
+    @Issue('https://github.com/gradle/gradle/issues/9904')
+    def "task with broken symlink in InputFiles is valid"() {
+        def inputFileTarget = file("brokenInputFileTarget")
+        def brokenInputFile = file('brokenInputFile').createLink(inputFileTarget)
+        def output = file("output.txt")
+
+        buildFile << """
+            class CustomTask extends DefaultTask {
+
+                @InputFiles FileCollection brokenInputFiles
+
+                @OutputFile File output
+
+                @TaskAction execute() {
+                    output.text = brokenInputFiles.files*.name
+                }
+            }
+            task inputBrokenLinkNameCollector(type: CustomTask) {
+                brokenInputFiles = files '${brokenInputFile}'
+                output = file '${output}'
+            }
+        """
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+        output.text == "[brokenInputFile]"
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+
+        when:
+        inputFileTarget.createFile()
+        run 'inputBrokenLinkNameCollector'
+        then:
+        executedAndNotSkipped ':inputBrokenLinkNameCollector'
+
+        when:
+        run 'inputBrokenLinkNameCollector'
+        then:
+        skipped ':inputBrokenLinkNameCollector'
+    }
+
+    // Validation fails with broken links assigned directly to `@InputFile`
```

Comment:
I am not sure these comments provide an additional value. Let's add a test for the behaviour (if we already don't have one) and remove the comment.

Reply:
Added tests for the two cases